### PR TITLE
Add network interface configuration schema for virtual machine network profile

### PR DESCRIPTION
**Pull request by basharg**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [ ] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  -  [ ] Ensure to copy the existing version into new directory structure for first commit (including refactoring) and then push new changes including version updates in separate commits. This is required to review the changes efficiently.
  - Adding a new service

- [ ] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 1]
```diff
@@ -9444,6 +9444,18 @@
...
+        "deleteOption": {
+          "type": "string",
+          "description": "Specify what happens to the network interface when the VM is deleted",
+          "enum": [
+            "Delete",
+            "Detach"
+          ],
+          "x-ms-enum": {
+            "name": "DeleteOptions",
+            "modelAsString": false
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9457]: We can sync with Jun about best practice, but my understanding is that modeling enums as strings is current best pratice to allow for future additios to the enum.

If not modeled as a string new enums reult in deserialization errors in old clients

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 2]
```diff
@@ -9462,17 +9474,424 @@
...
+      "properties": {
+        "domainNameLabel": {
+          "type": "string",
+          "description": "The Domain name label. The concatenation of the domain name label and vm [what] will be the domain name labels of the PublicIPAddress resources that will be created"
+        }
+      },
+      "required": [
+        "domainNameLabel"
+      ],
+      "description": "Describes a virtual machines network configuration's DNS settings."
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9500]: We should maybe indicate this is the domain name label prefix.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 3]
```diff
@@ -9462,17 +9474,424 @@
...
+        "publicIPAllocationMethod": {
+          "type": "string",
+          "description": "Specify the public IP allocation type",
+          "enum": [
+            "Dynamic",
+            "Static"
+          ],
+          "x-ms-enum": {
+            "name": "PublicIPAllocationMethod",
+            "modelAsString": false
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9558]: True for reasons above

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 4]
```diff
@@ -9462,17 +9474,424 @@
...
+          "type": "string",
+          "description": "Specify what happens to the public IP address when the VM is deleted",
+          "default": "Delete",
+          "enum": [
+            "Delete",
+            "Detach"
+          ],
+          "x-ms-enum": {
+            "name": "DeleteOptions",
+            "modelAsString": false
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9519]: model as string

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 5]
```diff
@@ -9462,17 +9474,424 @@
...
+      "properties": {
+        "idleTimeoutInMinutes": {
+          "type": "integer",
+          "format": "int32",
+          "description": "The idle timeout of the public IP address."
+        },
+        "deleteOption": {
+          "type": "string",
+          "description": "Specify what happens to the public IP address when the VM is deleted",
+          "default": "Delete",
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9512]: What do these defaults do?

Should this default to delete or null, what is best practice here?

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 6]
```diff
@@ -9462,17 +9474,424 @@
...
+          ],
+          "x-ms-enum": {
+            "name": "PublicIPAllocationMethod",
+            "modelAsString": false
+          }
+        }
+      },
+      "description": "Describes a virtual machines IP Configuration's PublicIPAddress configuration"
+    },
+    "LoadBalancerAndPublicIPSku": {
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9564]: How is this named in the VMSS schema for external customer consumption?

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 7]
```diff
@@ -9462,17 +9474,424 @@
...
+      "properties": {
+        "name": {
+          "type": "string",
+          "description": "Specify public IP sku name",
+          "enum": [
+            "Basic",
+            "Standard",
+            "Classic",
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9572]: Basic and Standard are the only ones allowed here.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 8]
```diff
@@ -9462,17 +9474,424 @@
...
+        "name": {
+          "type": "string",
+          "description": "The publicIP address configuration name."
+        },
+        "sku": {
+          "$ref": "#/definitions/LoadBalancerAndPublicIPSku"
+        },
+        "properties": {
+          "x-ms-client-flatten": true,
+          "$ref": "#/definitions/VirtualMachinePublicIPAddressConfigurationProperties"
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9609]: What does this mean?

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 9]
```diff
@@ -9462,17 +9474,424 @@
...
+        "tier": {
+          "type": "string",
+          "description": "Specify public IP sku tier",
+          "enum": [
+            "Regional",
+            "Global"
+          ],
+          "x-ms-enum": {
+            "name": "SkuTier",
+            "modelAsString": true
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9589]: Let's sync with Anh Cao about exposing this.

I am not sure if this is a released feature or not.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 10]
```diff
@@ -9462,17 +9474,424 @@
...
+      },
+      "required": [
+        "name"
+      ],
+      "description": "Describes a virtual machines IP Configuration's PublicIPAddress configuration"
+    },
+    "VirtualMachineNetworkInterfaceIPConfigurationProperties": {
+      "properties": {
+        "subnet": {
+          "$ref": "#/definitions/ApiEntityReference",
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9620]: How is this different from subresource?

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 11]
```diff
@@ -9462,17 +9474,424 @@
...
+          "maximum": 65535
+        },
+        "end": {
+          "type": "integer",
+          "minimum": 0,
+          "maximum": 65535
+        }
+      }
+    },
+    "DscpConfigurationProperties": {
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9724]: In the code this is a resource reference, not an inline set of properties.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 12]
```diff
@@ -9462,17 +9474,424 @@
...
+          "type": "string",
+          "default": "Delete",
+          "description": "Specify what happens to the network interface when the VM is deleted",
+          "enum": [
+            "Delete",
+            "Detach"
+          ],
+          "x-ms-enum": {
+            "name": "DeleteOptions",
+            "modelAsString": false
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9806]: True

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 13]
```diff
@@ -9462,17 +9474,424 @@
...
+    },
+    "VirtualMachineNetworkInterfaceConfigurationProperties": {
+      "properties": {
+        "primary": {
+          "type": "boolean",
+          "description": "Specifies the primary network interface in case the virtual machine has more than 1 network interface."
+        },
+        "deleteOption": {
+          "type": "string",
+          "default": "Delete",
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9798]: Should the default be delete or null here as above

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 14]
```diff
@@ -9462,17 +9474,424 @@
...
+          "x-ms-client-flatten": true,
+          "$ref": "#/definitions/VirtualMachineNetworkInterfaceIPConfigurationProperties"
+        }
+      },
+      "required": [
+        "name"
+      ],
+      "allOf": [
+        {
+          "$ref": "#/definitions/SubResource"
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9683]: What does this mean?

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 9 above)
Comment (szimmerMicrosoft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9589]: I was wrong, you should follow up in this thread:

Re: Public IP SKU assignment for VM NetworkInterfaceConfigurations

Seems Anavi is the SMe here.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 15]
```diff
@@ -9444,6 +9444,18 @@
         "primary": {
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9444]: Missing from this change is adding NetworkApiVersion to the VMSS netowrkProfile

Comment (basharg) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9444]: it is in VirtualMachineScaleSetNetworkProfile line 11459

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 2 above)
Comment (basharg) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9500]: should we rename the field then? and how should we describe given that it is concatenated with the profile configuration ID?

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 5 above)
Comment (basharg) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9512]: for VM with profile config, shouldn't the default be Delete?

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 6 above)
Comment (basharg) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9564]: VirtualMachineScaleSetPublicIPAddressConfiguration doesn't have a sku in swagger

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 5 above)
Comment (szimmerMicrosoft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9512]: The  default is delete on the server side.

In terms of client libraries I am not sure of the best practice, we should perhaps ask Adam and Jun

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 2 above)
Comment (szimmerMicrosoft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9500]: this is the name that is used in VMSS already, we want that compatibility.

Although there is different behavior.

We are also using "name" to mean the prefix as well (and in VMSS) so I think this is OK.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 16]
```diff
@@ -10810,6 +11229,19 @@
...
+          "type": "string",
+          "default": null,
+          "description": "Specify what happens to the network interface when the VM is deleted",
+          "enum": [
+            "Delete",
+            "Detach"
+          ],
+          "x-ms-enum": {
+            "name": "DeleteOptions",
+            "modelAsString": false
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:11243]: model as string true

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 17]
```diff
@@ -10849,6 +11281,19 @@
...
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:11295]: model as string true

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 10 above)
Comment (basharg) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9620]: diff seems to be that SubResource has "x-ms-azure-resource": true, while ApiEntity has better description. most resources are typed with subresource including external ones such as keyvault, only VMSS has some networking resources as ApiEntityReference

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 14 above)
Comment (basharg) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9683]: was copied from VirtualMachineScaleSetIPConfiguration, seems to imply that the object can also inherit properties from SubResource

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 6 above)
Comment (szimmerMicrosoft) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9564]: The naming in the Netwokring Docs is PublicIPAddressSku, PublicIPAddressSkuName, PublicIPAddressSkuTier

This seems reasonable to replicate.

Can we confirm that is the naming in the NRP swagger spec.
If so I think we should go with that?



https://docs.microsoft.com/en-us/rest/api/virtualnetwork/publicipaddresses/createorupdate#publicipaddresssku

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 18]
```diff
@@ -11011,6 +11360,17 @@
     },
     "VirtualMachineScaleSetNetworkProfile": {
       "properties": {
+        "networkApiVersion": {
```
Comment (szimmerMicrosoft) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:11363]: Need to add deleteOption for the publicIpadress config properties (Update And Create)

**Review by Sandido — APPROVED** [review]

**Comment by Sandido** [comment]
The other PR checks should conclude. We are not merging these PRs yet.

**Review by Sandido — CHANGES_REQUESTED** [review]
Please create or edit new Example files of the added functionality.

**Review by szimmer-dev — APPROVED** [review]
Looks good to me.

**Comment by Sandido** [comment]
@basharg , please address the .Net SDK CI failure.

**Comment by ArcturusZhang** [comment]
Breaking changes reported in this PR are all accumulated from previous PRs, therefore those are fine to ignore

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json [hunk 19]
```diff
@@ -9444,6 +9447,18 @@
...
+        },
+        "deleteOption": {
+          "type": "string",
+          "description": "Specify what happens to the network interface when the VM is deleted",
+          "enum": [
+            "Delete",
+            "Detach"
+          ],
+          "x-ms-enum": {
+            "name": "DeleteOptions",
```
Comment (Sandido) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9459]: Be aware that enums with the same name, even across multiple operations, must have the same values going forward at all times.

**Review by Sandido — APPROVED** [review]

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json (hunk 19 above)
Comment (Sandido) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/compute.json:9459]: @basharg

**Comment by msyyc** [comment]
The breaking changes for python are accumulated from previous PRs.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/examples/CreateAVmWithNetworkInterfaceConfiguration.json [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+          "osProfile": {
+            "adminUsername": "{your-username}",
+            "secrets": [],
+            "computerName": "myVM",
+            "windowsConfiguration": {
+              "provisionVMAgent": true,
+              "enableAutomaticUpdates": true
+            }
+          },
+          "networkProfile": {
```
Comment (majastrz) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2021-03-01/examples/CreateAVmWithNetworkInterfaceConfiguration.json:83]: >networkProfile [](start = 11, length = 14)

Why is the network profile completely different between the PUT request and PUT response?

**Review by yungezz — APPROVED** [review]

**Comment by yungezz** [comment]
hi @basharg, @Sandido is this PR ready to merge?

**Pull request merged**